contracts/deployables/warrant/WarrantBase.sol

```diff
@@ -0,0 +1,315 @@
+        return $.token;
+    }
+
+    /**
+     * @inheritdoc IWarrantBase
+     */
+    function feeToken() public view virtual override returns (address) {
+        WarrantBaseStorage storage $ = _getWarrantBaseStorage();
+        return $.feeToken;
+    }
+
+    /**
+     * @inheritdoc IWarrantBase
+     */
+    function tokenAmount() public view virtual override returns (uint256) {
+        WarrantBaseStorage storage $ = _getWarrantBaseStorage();
+        return $.tokenAmount;
+    }
+
+    /**
+     * @inheritdoc IWarrantBase
+     */
+    function tokenPrice() public view virtual override returns (uint256) {
+        WarrantBaseStorage storage $ = _getWarrantBaseStorage();
+        return $.tokenPrice;
+    }
+
+    /**
+     * @inheritdoc IWarrantBase
+     */
+    function feeReceiver() public view virtual override returns (address) {
+        WarrantBaseStorage storage $ = _getWarrantBaseStorage();
+        return $.feeReceiver;
+    }
+
+    /**
+     * @inheritdoc IWarrantBase
+     */
+    function expiration() public view virtual override returns (uint256) {
+        WarrantBaseStorage storage $ = _getWarrantBaseStorage();
+        return $.expiration;
+    }
+
+    /**
+     * @inheritdoc IWarrantBase
+     */
+    function executed() public view virtual override returns (bool) {
+        WarrantBaseStorage storage $ = _getWarrantBaseStorage();
+        return $.executed;
+    }
+
+    // --- State-Changing Functions ---
+
+    /**
+     * @inheritdoc IWarrantBase
+     * @dev Validates conditions, collects fee, and delegates to _executeWarrant
+     */
+    function execute(address recipient_) public virtual override {
+        WarrantBaseStorage storage $ = _getWarrantBaseStorage();
+
+        // Validate caller and state
+        if (msg.sender != $.warrantHolder) revert OnlyWarrantHolder();
+        if (recipient_ == address(0)) revert AddressZero();
+        if ($.executed) revert AlreadyExecuted();
+
+        // Check expiration based on time mode and token lock if relative time
+        if ($.relativeTime) {
+            if (IVotesERC20V1($.token).locked()) revert TokenLocked();
+            if (
+                block.timestamp >
+                IVotesERC20V1($.token).getUnlockTime() + $.expiration
+            ) revert Expired();
+        } else {
+            if (block.timestamp > $.expiration) revert Expired();
+        }
+
+        // Calculate and collect fee
+        uint256 feeAmount = ($.tokenAmount * $.tokenPrice) / PRECISION;
+        IERC20($.feeToken).safeTransferFrom(
+            msg.sender,
+            $.feeReceiver,
+            feeAmount
+        );
+
+        // Mark as executed before external call
+        $.executed = true;
+
+        // Delegate to implementation-specific logic
+        _executeWarrant(recipient_);
+
+        emit Executed(recipient_);
+    }
+
+    /**
+     * @inheritdoc IWarrantBase
+     */
+    function clawback(address recipient_) public virtual override onlyOwner {
```

Comment:
What if the owner calls `clawback()` twice? I assume the `IERC20($.token).safeTransfer(recipient_, $.tokenAmount);` line will fail? Is it safe to rely on that?

Reply:
Yes, that safeTransfer line will fail, and it's definitely safe to rely on that. We could add a check for it so we could throw a more descriptive error, but in my opinion its not necessary.

Reply:
The expected flow is that the Countersign contract would have a list of TXs to execute, one of which would be deploying this Warrant contract, and a followup TX which transfers the correct token amount to this contract.

Currently, if this contract ends up with more `token` than `tokenAmount`, those tokens will be locked in the contract. I remember discussing this with Eddie and John and we decided this was fine, but am open to counter arguments on it.

Reply:
if this was discussed and decided upon already, i'm not going to argue against it. just wanted to make sure it _has_ been discussed